optaweb-vehicle-routing-frontend/src/components/SearchBox.tsx

```diff
@@ -0,0 +1,157 @@
+import { Button, Text, TextContent, TextInput, TextVariants } from '@patternfly/react-core';
+import { PlusSquareIcon } from '@patternfly/react-icons';
+import { OpenStreetMapProvider } from 'leaflet-geosearch';
+import * as React from 'react';
+import { ILatLng } from '../store/route/types';
+
+export interface IProps {
+  searchDelay: number;
+  addHandler: (result: IResult) => void;
+}
+
+export interface IState {
+  query: string;
+  results: IResult[];
+  attributions: string[];
+}
+
+export interface IResult {
+  address: string;
+  latLng: ILatLng;
+}
+
+class SearchBox extends React.Component<IProps, IState> {
+
+  static defaultProps = {
+    searchDelay: 500,
+  };
+
+  private searchProvider = new OpenStreetMapProvider({ params: { countrycodes: 'BE' } });
+  private timeoutId: number;
+
+  constructor(props: IProps) {
+    super(props);
+
+    this.state = {
+      attributions: [],
+      query: '',
+      results: [],
+    };
+
+    this.handleTextInputChange = this.handleTextInputChange.bind(this);
+    this.handleClick = this.handleClick.bind(this);
+  }
+
+  componentWillUnmount() {
+    window.clearTimeout(this.timeoutId);
+  }
+
+  handleTextInputChange(query: string): void {
+    window.clearTimeout(this.timeoutId);
+    if (query.trim() !== '') {
+      this.timeoutId = window.setTimeout(
+        async () => {
+          const searchResults = await this.searchProvider.search({ query });
+          if (this.state.query !== query) {
+            return;
+          }
+          this.setState({
+            attributions: searchResults
+              .map(result => result.raw.licence)
+              // filter out duplicate elements
+              .filter((value, index, array) => array.indexOf(value) === index),
+            results: searchResults.map(result => ({
+              address: result.label,
+              latLng: { lat: result.y, lng: result.x },
+            })),
+          });
+        },
+        this.props.searchDelay);
+    } else {
+      this.setState({ results: [], attributions: [] });
+    }
+    this.setState({ query });
+  }
+
+  handleClick(index: number) {
+    this.props.addHandler(this.state.results[index]);
+    this.setState({
+      attributions: [],
+      query: '',
+      results: [],
+    });
+  }
+
+  render(): React.ReactNode {
+    const { attributions, query, results } = this.state;
+    return (
+      <div>
+        <TextInput
+          value={query}
+          type="search"
+          placeholder={'Search to add a location...'}
+          aria-label="geosearch text input"
+          onChange={this.handleTextInputChange}
+        />
+        {results.length > 0 &&
+        <div className="pf-c-options-menu pf-m-expanded">
+          <ul className="pf-c-options-menu__menu">
+            {results.map((result, index) => (
+              <li key={`result${index}`}>
+                <div className="pf-c-options-menu__menu-item">
+                  {result.address}
+                  <Button
```

Comment:
A soft suggestion - instead of creating the "+" button for each suggested place, you could change the "text" cursor into either "pointer" or "default" (http://www.javascripter.net/faq/stylesc.htm) and accept just a click on the list item.

Reply:
Good suggestion but it's premature. The widget will appear soon in a completely new screen so I'd rather leave this fine tuning until after the new context is more clear. I also want to be careful with unnecessary customization of Patternfly widgets as that means divergence from what should be there ready to use. I'd A) look for a better widget that's available, or B) make (and help) them add the widgets we need.